src/ToolingSharedLibrary/Includes/VbsEnclaveABI/Shared/ConversionHelpers.h

```
@@ -0,0 +1,543 @@
+        {
+            if (!wstr)
+            {
+                return {};
+            }
+
+            return std::wstring(wstr->wchars.begin(), wstr->wchars.end());
+        }
+        else
+        {
+            return std::wstring(wstr.wchars.begin(), wstr.wchars.end());
+        }
+    }
+
+    template<typename FlatbufferType>
+    inline FlatbufferType CreateWStringT(const std::wstring& wchars)
+    {
+        using Decayed = std::decay_t<FlatbufferType>;
+        if constexpr (UniquePtr<Decayed>)
+        {
+            using InnerType = unique_ptr_inner_type_t<Decayed>;
+            auto ptr = std::make_unique<InnerType>();
+            THROW_IF_NULL_ALLOC(ptr);
+            ptr->wchars.assign(wchars.begin(), wchars.end());
+            return ptr;
+        }
+        else if constexpr (NonWrappedStruct<Decayed>)
+        {
+            Decayed wcharT {};
+            wcharT.wchars.assign(wchars.begin(), wchars.end());
+            return wcharT;
+        }
+        else
+        {
+            static_assert(always_false<FlatbufferType>::value, "CreateWStringT: Unsupported FlatbufferType (must be unique_ptr or struct)");
+        }
+    }
+
+    template <typename SrcContainer, typename TargetContainer, typename ConvertFunc>
+    TargetContainer TransformContainer(const SrcContainer& src_container, ConvertFunc&& conversion_func)
+    {
+        TargetContainer target_container;
+        target_container.reserve(src_container.size());
+        for (const auto& value : src_container)
+        {
+            target_container.emplace_back(conversion_func(value));
+        }
+
+        return target_container;
+    }
+
+    template <typename SrcContainer, typename ArrayContainer, typename ConvertFunc>
+    ArrayContainer TransformToStdArray(const SrcContainer& src_container, ConvertFunc&& conversion_func)
+    {
+        ArrayContainer array_container {};
+        size_t N = std::min(array_container.size(), src_container.size());
```

this would be lossy i think?  if dest is too small

---

Would it? We choose the minimum out of both so the array_container if it is small would win out.

---

(I was find with the "container" naming)

[Round-tripability]

I'm still not convinced - if we're doing DT(std::vector 2 elems) -> FBT (std::array 1 elem) -> DT(std::vector), you're going to lose an element after the round-trip, aren't you?

---

So flatbuffer doesn't support std::arrays. It only supports std::vectors.  so src in this case would end up being a vector whose size we won't know at compile time. 

TLDR: used fail fast instead and made sure our codegen only passes SrcContainers that are the same size as the array.

I realized what I was using the std::min() for was really just the fact that there is a single case where the vector size != the array size. That is only when there is an out param std::array. We only ever call the ConvertType for each in/inout parameter  in a stub function, placing the result in the appropriate flatbuffer struct field. For out parameters, that would mean the vector size would be zero.

So what I did was make sure that our param std::arrays also get their value updated via the ConvertType function. After that the size of both are the same. See the comment on line 416 in CppCodeBuilder.cpp.

---

great - i'm glad we could add this ratchet.  and with it remove the `#undefs` a the top.  ngl - I'm still trying to grok the std::array out param case

---

nvm i get it now - invariant must hold, even in the zero case - nice